### ISSUE TITLE
Durable state for multi-host fault tolerance

### ISSUE THREAD

**Issue by blp** [feldera/feldera#658]
At a very high level, the requirements for durable state for multi-host fault tolerance are like those for single-host fault tolerance.

At a more realistic level, implementing the [single-host specification](https://github.com/feldera/feldera/issues/657) would require simultaneously committing databases across all the hosts, which is difficult and common embedded databases such as `rocksdb` do not support it.

Thus, to support multihost fault tolerance, we instead extend our database to support multiple versions (two versions is enough), so that we can write and commit a new version on all our hosts and then fall back to the earlier version if any of the commits fail. Here is a copy of the [specification](https://github.com/feldera/dist-design#database) from the distributed design document:

Let `Version` be a `u64` type representing a database version serial number:

  * A new database has only one version, numbered 0, which is empty.

  * Opening a database at version `v` causes the other version, if  any, to be discarded.

  * Committing a database that was opened at version `v` creates a new    version `v + 1`, either of which may be opened with `open`.

  * A database closed in any way other than `commit` retains the    version that was passed to `open`.  If a crash happens during    `commit` then it's OK for both versions or just the version passed    to `open` to be available on recovery.

A database implementation presents the following interface to workers:

* `fn versions() -> Range<Version>`

  Reports the versions that the database can open.  The returned  `Range` has length 1 or 2.

* `fn open(v: Version) -> Result<Database>`

  Opens the database.  Reads and writes will occur in version `v + 1`,  which is initially a copy of version `v`.  The database discards  versions other than `v`.

* `fn commit(db: Database) -> Result<()>`

  Commits the current version of `db` and closes the database.  After  a successful return, if `v` was the version passed to `open()`, then  versions `v` and `v + 1` may now be opened, where `v` has the same  contents as before and `v + 1` reflects any changes made while the  database was open.

**Comment by ryzhyk** [comment]
I wonder if we can take advantage of already existing machinery. Virtually all the state in the database consists of traces.  Each entry in a trace has an associated timestamp, which I think is equivalent to step number. Today we use a (very) lossy representation for top-level timestamps as `()`. This representation is not suitable for persistent storage and for the distributed runtime, so we will anyway have to switch to integer timestamps.  Can these timestamps be used to simulate a multi-version database by truncating values with timestamps that represent changes made after the last commit?

**Comment by blp** [comment]
> I wonder if we can take advantage of already existing machinery. Virtually all the state in the database consists of traces. Each entry in a trace has an associated timestamp, which I think is equivalent to step number. Today we use a (very) lossy representation for top-level timestamps as `()`. This representation is not suitable for persistent storage and for the distributed runtime, so we will anyway have to switch to integer timestamps. Can these timestamps be used to simulate a multi-version database by truncating values with timestamps that represent changes made after the last commit?

That's a great idea, thank you.

Suppose that takes care of recording multiple versions. We have another question in the design space. When we want to open an older version, how do we do it? Some possibilities:
* Find entries that contain newer versions and delete the new ones. This could be done by keeping a list of entries that have been changed, by using a RocksDB [transaction log iterator](https://github.com/facebook/rocksdb/blob/main/include/rocksdb/transaction_log.h), by iterating through the entire database, or some other clever way.
* Use some kind of virtual version numbering and mapping, so that whenever we read a record with obsolete data we ignore it, and whenever we write or compact a record we discard any obsolete data.

These questions also come up without that optimization, of course. I wrote down [some ideas](https://github.com/feldera/dist-design/blob/main/rocksdb.md) earlier, too.

**Comment by ryzhyk** [comment]
Right, ideally the storage engine would allow pruning state beyond the opened timestamp efficiently.  The trace data structure was designed for this (although we are not currently using this functionality).  Not sure about RocksDB, but it sounds like @blp has some ideas about this.

This is also related to the concept of compacting timestamps.  Once the state has been fully checkpointed, we no longer need to distinguish the timestamps in the checkpointed state. We could just go and change them all to 0, which would reduce the size of the trace (by merging multiple timestamps into one and removing entries whose resulting weight after merging is 0) and speed up access, but that'd be a very expensive operation. Instead we can do this lazily when merging layers in the trace or in the LSM tree in case of RocksDB.  Again, the trace data structure was designed for this.

**Comment by ryzhyk** [comment]
To clarify my previous point about the trace data structure supporting timestamp truncation:
- Each layer in the trace has a range of timestamps that occur in this layer.
- These ranges decrease monotonically when moving from smaller to large layers (i.e., smaller layers contain more recent timestamps).
- The pruning algorithm would:
  -  Remove layers whose timestamp range is more recent than the version we're opening (i.e., the earliest timestamp in the range is more recent than the version)
  - Don't touch layers whose latest timestamp is smaller than version
  - Process all other layers to remove timesamps greates than version (I guess this can be further optimized by somehow doing it lazily).
